storage/pebble/my_receipts.go

```diff
@@ -17,29 +19,63 @@ import (
 // The wrapper adds the ability to "MY execution receipt", from the viewpoint
 // of an individual Execution Node.
 type MyExecutionReceipts struct {
-	genericReceipts *ExecutionReceipts
-	db              *pebble.DB
-	cache           *Cache[flow.Identifier, *flow.ExecutionReceipt]
+	genericReceipts     *ExecutionReceipts
+	db                  *pebble.DB
+	cache               *Cache[flow.Identifier, *flow.ExecutionReceipt]
+	indexingOwnReceipts sync.Mutex // lock to ensure only one receipt is stored per block
 }
 
 // NewMyExecutionReceipts creates instance of MyExecutionReceipts which is a wrapper wrapper around pebble.ExecutionReceipts
 // It's useful for execution nodes to keep track of produced execution receipts.
 func NewMyExecutionReceipts(collector module.CacheMetrics, db *pebble.DB, receipts *ExecutionReceipts) *MyExecutionReceipts {
+
+	mr := &MyExecutionReceipts{
+		genericReceipts: receipts,
+		db:              db,
+	}
+
 	store := func(key flow.Identifier, receipt *flow.ExecutionReceipt) func(storage.PebbleReaderBatchWriter) error {
 		// assemble DB operations to store receipt (no execution)
 		storeReceiptOps := receipts.storeTx(receipt)
 		// assemble DB operations to index receipt as one of my own (no execution)
 		blockID := receipt.ExecutionResult.BlockID
 		receiptID := receipt.ID()
+
+		// check if the block already has a receipt stored
 		indexOwnReceiptOps := operation.IndexOwnExecutionReceipt(blockID, receiptID)
 
 		return func(rw storage.PebbleReaderBatchWriter) error {
+
 			err := storeReceiptOps(rw) // execute operations to store receipt
 			if err != nil {
 				return fmt.Errorf("could not store receipt: %w", err)
 			}
 
-			_, w := rw.ReaderWriter()
+			r, w := rw.ReaderWriter()
+
+			// acquiring the lock is necessary to avoid dirty reads when calling LookupOwnExecutionReceipt
+			mr.indexingOwnReceipts.Lock()
+			rw.AddCallback(func() {
```

Comment:
are callbacks always called, or only when successful?

Reply:
callbacks are always called regardless the batch updates fails or not.